lib/Cake/Network/CakeResponse.php

```diff
@@ -1464,7 +1465,7 @@ protected function _sendFile($file, $range) {
 		$file->open('rb');
 
 		$end = $start = false;
-		if ($range) {
+		if (!empty($range)) {
```

Comment:
the previous was 
a) more correct as it doesn't cloak
b) otherwise completely identical

please see cakephp coding guidlines for this.

Reply:
> The expression $range of type array is implicitly converted to a boolean; are you sure this is intended? If so, consider using ! empty($expr) instead to make it clear that you intend to check for an array without elements. 

According to CakePHP [coding standards](https://book.cakephp.org/2.0/en/contributing/cakephp-coding-conventions.html#comparison) regarding comparison:
> Always try to be as strict as possible.

Reply:
You didnt fix strictness here, you just did the opposite. You cloaked a var, which is bad.
Read the part about not using empty() and isset() wrongly.

Reply:
Could you be more specific in referencing to the guidelines? I have not found any mention of `empty` on that page. I do not fully understand how making the intention of the code clearer makes it worse.

Reply:
It is pretty highlighted on the page actually:
https://book.cakephp.org/3.0/en/contributing/cakephp-coding-conventions.html#careful-when-using-empty-isset

You may also read [this](https://www.dereuromark.de/2016/06/01/use-isset-and-empty-wisely/).

Reply:
> When working with variables, it is better to rely on type-coercion to boolean instead of empty().

That statement is highly arguable. In my opinion it is quite confusing and is in fact buggy when a string or an array is converted to a boolean for comparison that way. 

In this particular case `list()` that is called below expects `$range` to be an array. If `$range` is `true` or `1` (and this may happen if `true` or `1` is passed as the parameter of `_sendFile` method) this code makes no sense:
```
list($start, $end) = true;
```
The values of `$start` and `$end` are `null` in this case. 
```
$end = $start = false;
var_export($start); // outputs false
var_export($end);   // outputs false
list($start, $end) = true;
var_export($start); // outputs NULL
var_export($end);   // outputs NULL
```
And that seems to be not what is expected here. The code inside `if ($start !== false) {` below would not be executed.

So the correct check would be `if (is_array($range) && !empty($range)) {` here.

Reply:
Let's please stop that discussion.
You are wrong here in terms of functionality. It does not make a difference, those statements (before and after your change) are identical. The downside of cloaking is explained in both linked resources. This has nothing to do with casting.
Please revert this and focus on the actual changes that matter.

Reply:
I've got an impression that you did not fully read my previous post :(.

Reply:
I did - but you are then not understanding how PHP works. There is no casting happening.
The $range is still what it is, even after such a check. http://sandbox.onlinephpfunctions.com/code/580d6dbc977dd40112d36dcb8976662cc9d6e0bf

Reply:
I understand that no casting happens and the value of `$range` does not change. That's why I wrote "and this may happen if true or 1 is passed as the parameter of _sendFile method". The values of `$start` and `$end` can be null in some scenarios. http://sandbox.onlinephpfunctions.com/code/6623b50ebc865963924eb4fa8730c38bd2d4d922

Reply:
Is there a test case to proof this edge case?
Otherwise the change below looks too dangerous.

Reply:
If there would be a unit test for this scenario the bug would not be in the code. Why the code in http://sandbox.onlinephpfunctions.com/code/6623b50ebc865963924eb4fa8730c38bd2d4d922 is not convincing?

Reply:
Maybe [this](http://sandbox.onlinephpfunctions.com/code/989185d2a0747f4cf2dbc15714b567d1b3a5ccd1) would be more convincing?